Extract reusable part of Webpack config and put in @wordpress/scripts

gziolo opened the pull request:
## Description
Replaces #12685.

This PR has been started by @griffbrad in #1268 and was desrbibed as follows (with some updates from myself):

> It adds a new `@wordpress/webpack-config` package, which pulls the base webpack config into a package so it can be easily re-used by 3rd-party developers. Once merged, this would allow developers to either depend on the package or run `wp-scripts build` from `@wordpress/scripts` to build blocks and extensions using this config.
>
> By having something like this in core, we'd have some simple build tools available out of the box so that block developers don't need to hunt down 3rd-party tools or configure webpack/babel manually.  Further, the `@wordpress/webpack-config` package may also be useful for 3rd-party tool developers (e.g. create-guten-block, gutenblock, etc) who want to better ensure their compatibility with Gutenberg. 

## Examples of Webpack configs used by plugins

* [Gutenberg SlotFill and Filter demos](https://github.com/ryanwelcher/slotfill-and-filter-demos/blob/master/webpack.config.js) by @ryanwelcher 
* [Juxtapose Block](https://github.com/mkaz/juxtapose-block/blob/master/webpack.config.js) by @mkaz 
* [Gutenberg Custom Fields](https://github.com/youknowriad/gcf/blob/master/webpack.config.js) by @youknowriad 
* [Gutenberg Extensibility Workshop](https://github.com/youknowriad/gutenberg-extensibility-workshop/blob/master/webpack.config.js) by @youknowriad 
* [Drop it](https://github.com/youknowriad/dropit/blob/master/webpack.config.js) by @youknowriad 
* [Tomodomo Block Starter](https://github.com/TomodomoCo/wp-block-starter/blob/master/webpack.config.js) by @chrisvanpatten 

## How has this been tested?

The first thing to test with this PR is just a normal Gutenberg build because much of the webpack config has been extracted into a package.  From what I can tell, everything still builds correctly, tests pass, and Gutenberg appears to function as expected.
 
Testing block building using `wp-scripts` is a bit tricky because you're running scripts from unpublished packages only available locally.  To make this work, I do the following:

1. Build Gutenberg itself on this feature branch with `npm run build`.
2. Clone the following example plugin into your plugins folder: https://github.com/griffbrad/wp-scripts-build-block-example-plugin
3. In that plugin's folder, run `npm run copy-local-gutenberg-packages && npm i`.
4. Run `npm run build` in the example plugin's folder.  The block's code in `src/index.js` will be built with the output being placed in `build`.
5. Activate the example plugin and ensure the "WP Scripts Build Tool" block is available and functioning.

These testing instructions copy the built Gutenberg plugin and its packages into the example plugin itself because I've found that using a file ref to traverse back to the parent directory causes NPM to randomly crash and spew non-sense.  

## Types of changes
New feature

## Checklist:
- [x] My code is tested. (Working on a script that does some basic end-to-end testing of the process.)
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

## Planned follow-up tasks
- [x] Simpler default for webpack output configuration.
- [ ] Docs with examples for modifying the base webpack config when needed.  Mostly thinking of this as a "zero configuration" approach, but I can foresee some common tweaks we may want to support.

lynnbgriffith committed:
New build-config package with webpack config.

Pull the Gutenberg webpack config into a package so it can be re-used for
block/extension development.

lynnbgriffith committed:
Require new build-config package.

lynnbgriffith committed:
Dynamically handle WP externals with a function.

Use code from WP Calypso for handling WP externals so we don't have to have the
actual list of packages accessible in our webpack configuration.

lynnbgriffith committed:
Use webpack config from build-config package.

lynnbgriffith committed:
Require build-config package.

lynnbgriffith committed:
Adjust file refs for WP packages.

lynnbgriffith committed:
Move main gutenberg entry definition and webpack copy plugin out of build-config.

lynnbgriffith committed:
Add react-dev-utils for formatting webpack compiler messages.

lynnbgriffith committed:
Implement build script using webpack config from build-config.

lynnbgriffith committed:
Adjust output path so build goes to working directory.

gziolo committed:
Update package name to webpack-config

gziolo committed:
Apply more tweaks to the way webpack config package is structured

gziolo commented:
This PR could be merged as is and we could tackle zero-config support for `wp-scripts build` and `wp-scripts start` separately.


@youknowriad raised some concerns on WordPress Slack (link requires registration at https://make.wordpress.org/chat/) https://wordpress.slack.com/archives/C5UNMSU4R/p1549962381008100. The downside of exposing a shared webpack config to npm is that we will have to maintain its public API. This needs to be discussed further as I don't think that putting it inside `@wordpress/scripts` changes as much because it's going to be still possible to reference it from npm. However, this will give more clear signal that it isn't part of the public API.

mkaz commented:
My initial test following the instructions:

```
> npm run build

> @ build /home/mkaz/src/wp/wp-scripts-build-block-example-plugin
> wp-scripts build

Webpack config file is missing.
```

Created a webpack.config.js using from webpack-config package:
```
const { config } = require( '@wordpress/webpack-config' );

module.exports = Object.assign( {}, config, {
	// apply your changes here
} );
```
Errors with: `Error: Cannot find module '@wordpress/webpack-config'`

Added to dependencies in package.json:
```
"@wordpress/webpack-config": "file:./gutenberg-local-packages/packages/webpack-config"
```

Error with JSX syntax:

```
> npm run build

> @ build /home/mkaz/src/wp/wp-scripts-build-block-example-plugin
> wp-scripts build

Hash: a6c3b571764ff714d575
Version: webpack 4.8.3
Time: 262ms
Built at: 02/12/2019 6:33:01 AM
 1 asset
Entrypoint main = main.js
[0] ./src/index.js 6.22 KiB {0} [built] [failed] [1 error]

ERROR in ./src/index.js
Module build failed: SyntaxError: /home/mkaz/src/wp/wp-scripts-build-block-example-plugin/src/index.js: Unexpected token (14:19)

  12 |         },
  13 |         edit: function () {
> 14 |             return <div>Hey, Gutenberg!</div>;
     |                    ^
```

gziolo commented:
@mkaz thanks for testing 🙇  I think that you will temporarily need to provide your own Babel file. I haven't tested it with external projects. I'm validating whether it produces the same output for Gutenberg core :)

mkaz commented:
Adding a `.babelrc` and installing babel-plugin-transform-react-jsx, and babel-preset-env
gets me to the next error which my first initial tries did not get past.

```
> npm run build

> @ build /home/mkaz/src/wp/wp-scripts-build-block-example-plugin
> wp-scripts build

Hash: d42ad38e5514fb4f4830
Version: webpack 4.8.3
Time: 265ms
Built at: 02/12/2019 6:46:41 AM
 1 asset
Entrypoint main = main.js
[0] ./src/index.js 2.52 KiB {0} [built] [failed] [1 error]

ERROR in ./src/index.js
Module build failed: TypeError: Cannot read property 'bindings' of null
    at Scope.moveBindingTo (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/scope/index.js:867:13)
    at BlockScoping.updateScopeInfo (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/babel-plugin-transform-es2015-block-scoping/lib/index.js:364:17)
    at BlockScoping.run (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/babel-plugin-transform-es2015-block-scoping/lib/index.js:330:12)
    at PluginPass.BlockStatementSwitchStatementProgram (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/babel-plugin-transform-es2015-block-scoping/lib/index.js:70:24)
    at newFn (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/visitors.js:193:21)
    at NodePath._call (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/path/context.js:40:17)
    at NodePath.visit (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/path/context.js:88:12)
    at TraversalContext.visitQueue (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/context.js:118:16)
    at TraversalContext.visitSingle (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/context.js:90:19)
    at TraversalContext.visit (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/context.js:146:19)
    at Function.traverse.node (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/index.js:94:17)
    at traverse (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/traverse/lib/index.js:76:12)
    at transformFile (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/core/lib/transformation/index.js:88:29)
    at runSync (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/core/lib/transformation/index.js:45:3)
    at runAsync (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/core/lib/transformation/index.js:35:14)
    at process.nextTick (/home/mkaz/src/wp/wp-scripts-build-block-example-plugin/node_modules/@babel/core/lib/transform.js:34:34)
    at process._tickCallback (internal/process/next_tick.js:61:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! @ build: `wp-scripts build`
npm ERR! Exit status 2
npm ERR! 
```

gziolo commented:
The way we handle externals has changed to avoid using the whitelist of packages. Instead, we use a pattern match. It's awesome and very expected as this is going to be bulletproof for external plugins and themes. However, it would be great to confirm that this subtle difference in the transipiled code is expected:

_Before_:

```js
/***/ (function(module, exports) {

(function() { module.exports = this["wp"]["domReady"]; }());

/***/ })
```

_After_:
/***/ (function(module, exports) {

module.exports = wp.domReady;

/***/ })
```js

It's less code so I would assume it's better. I'm not quite sure what was the original reasoning behind the used syntax.

simison commented on packages/webpack-config/config.js:
```diff
@@ -0,0 +1,103 @@
+		moment: 'moment',
+		jquery: 'jQuery',
+		lodash: 'lodash',
+		'lodash-es': 'lodash',
```
If the config is meant to be used for other parts of admin interface, outside the block editor, should this list contain everything WP can externalise?

https://developer.wordpress.org/reference/functions/wp_enqueue_script/#default-scripts-included-and-registered-by-wordpress

gziolo replied:
I'm not sure. For the time being, it's editor specific and I hope that the number of dependencies won't increase here.

gziolo committed:
Update the way externals are handled

gziolo commented:
I managed to fix the generated output mentioned in my previous comment. With ac86224 this is how the output looks like:

```js
/***/ (function(module, exports) {

(function() { module.exports = this["wp"]["domReady"]; }());

/***/ })
```

gziolo commented:
> Adding a `.babelrc` and installing babel-plugin-transform-react-jsx, and babel-preset-env
> gets me to the next error which my first initial tries did not get past.

@mkaz did you add `entry` and `output` fields to the config files? I think they are missing at the moment. I'm not quite sure whether we should provide some defaults and what would that be.

gziolo commented:
Let's move to 5.2 milestone as it still needs some polishing.

gziolo committed:
Add default values for entry and output

nosolosw commented:
I've tested this with two external scenarios:

- a plugin wants to use the `webpack-config` package directly. Example at https://github.com/nosolosw/understanding-gutenberg/pull/5
- a plugin wants to use this with `wp-scripts`. Example at https://github.com/nosolosw/understanding-gutenberg/pull/6

I've listed the result of the experiment in the PRs above.

As far I've tested, if we provide the webpack config as a template to create a new config, the webpack-cli will lookup every library/util (babel-loader, source-map-loader) under the consumer current working directory (my example plugin). So those packages need to be installed by the consumer.

I understand we want to do this to abstract away the details of the babel/webpack setup (that is: config + packages), so perhaps what we need to do is to provide a webpack default through the `@wordpress/scripts` if none is provided. That way, 3rd parties can bootstrap their plugins with zero-config. They'll also have the ability to provide a webpack config if they want, and, in that case, they'll need to provide the utils they use (babel-loader, etc).

gziolo commented:
> I understand we want to do this to abstract away the details of the babel/webpack setup (that is: config + packages), so perhaps what we need to do is to provide a webpack default through the `@wordpress/scripts` if none is provided. That way, 3rd parties can bootstrap their plugins with zero-config. They'll also have the ability to provide a webpack config if they want, and, in that case, they'll need to provide the utils they use (babel-loader, etc).

I share the same sentiment after digging a bit deeper into the scope of what could be shared. It also align with the concerns @youknowriad raised. To make it simpler for starts let's not create a new package with webpack config only but optimize for the most commons use case. I'll update PR shortly according to what was described by @nosolosw.

gziolo committed:
Move shared webpack config under @wordpress/scripts package

gziolo commented:
I pushed changes, it should be ready to go, assuming it works properly with external projects.
I tested it as follows:
• added `src/index.js`
• removed content of `webpack.config.js` and re-exported `@wordpress/scripts/config/webpack.config.js` instead
* `npm run dev` & `npm run build` - in both cases I could see `build/index.js` updated

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,118 @@
+		path: path.resolve( process.cwd(), 'build' ),
+	},
+	externals,
+	resolve: {
```
Due to the way local packages work I can't test this properly with an external plugin. https://github.com/nosolosw/understanding-gutenberg/pull/6 doesn't work because `@wordpress/scripts` dependencies (being local) aren't hoisted to the top-level `node_modules` folder, but are installed in `node_modules/@wordpress/scripts/node_modules` instead. If I could use `@wordpress/scripts` from npm the dependencies would be hoisted and the example would work.

**There is a gotcha with using `pwd.cwd` here, though.**

As far as I understand, this tells webpack to lookup dependencies (`babel-loader`, `source-map-loader`) in the `node_modules` folder of the consumer (the plugin/app/etc that uses this config). Due to how npm works, we can't guarantee that the versions in the top-level `node_modules` of the consumer are the ones we expect. Example:

- a package other than `@wordpress/scripts` uses a different version of webpack/babel-loader/etc and it gets installed first. The top-level `node_modules/` will contain those versions.
- `@wordpress/scripts` is installed after, and because the top-level already has `webpack`, `babel-loader`, etc, its dependencies are installed under `node_module/@wordpress/scripts/node_modules`.

When webpack goes to find the dependencies, they won't be the ones we want. I don't know how frequent this can be. If the user only pulls `@wordpress/packages` as dependencies, this won't be a problem. But as soon as they start using external packages, it will.

**How to solve this?**

https://github.com/WordPress/gutenberg/pull/13877 is a potential approach. It doesn't use `pwd.cwd` but introduces some changes to path resolution and a plugin to teach Webpack how to behave.

We could use that as a base, merge, and then create a second PR on top of it that includes the changes in this PR.

nosolosw replied:
This is no longer a concern, as we figured we can use `require.resolve` for path resolution of the utils webpack uses. When the file is loaded, the paths will be resolved to point to `@wordpress/scripts`'s `node_modules` directory instead of the consumer.

gziolo committed:
Improve the way how loaders are handled

gziolo commented:
After some discussion with @nosolosw, I think we figured out how to move forward with this PR. We revisited the way how modules and loaders, in particular, are handled with 1e0ff59.

We will still need #13877 to add the default Babel config for Webpack when it is not provided, set this config as a default one and update the corresponding docs.

aduth commented on packages/scripts/utils/string.js:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @return {string} Camel-cased string.
+ */
+function camelCaseDash( string ) {
```
If we're building something new, I'm wondering if we couldn't use this as an opportunity to address the fact that this function is insufficient in representing our capitalization standards, resulting in global scoped variables like `wp.escapeHtml` which should be corrected as `wp.escapeHTML`.

https://wordpress.slack.com/archives/C5UNMSU4R/p1542385406412600
https://wordpress.slack.com/archives/C5UNMSU4R/p1542722527426400

gziolo replied:
I extracted it from the existing webpack config. It's still "private" but I'm happy to iterate on it and maybe even put it into a separate package.

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+			{
+				test: /\.js$/,
+				exclude: [
+					/block-serialization-spec-parser/,
```
`block-serialization-spec-parser` and `is-shallow-equal` are Gutenberg packages. These exclusions should be added by the Gutenberg's Webpack config.

nosolosw replied:
We want to be able to extend these exclusions (perhaps via an env variable?). Some plugins will want to add to these as well, so they can reuse the same mechanism we come up for Gutenberg.

gziolo replied:
Yes, I didn’t have a good idea how to tackle it to be honest. I will think about it tomorrow.

gziolo replied:
It looks like it is no longer necessary. I have no idea why I added it in the first place. Maybe it was a bug in weback which was resolved 🤷‍♂️

gziolo replied:
It wasn't that easy but I fixed it with https://github.com/WordPress/gutenberg/pull/13814/commits/3a8093b31eedb28cbf7b1f35944cbb97012adc2a.

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+	plugins: [
+		// GUTENBERG_BUNDLE_ANALYZER global variable enables utility that represents bundle content
+		// as convenient interactive zoomable treemap.
+		process.env.GUTENBERG_BUNDLE_ANALYZER && new BundleAnalyzerPlugin(),
```
I don't see this used anywhere. I'd suggest renaming to:

```suggestion
		process.env.BUNDLE_ANALYZER && new BundleAnalyzerPlugin(),
```

nosolosw replied:
For a future iteration: it'd be great to have a `wp-scripts analyze` command that sets up the env variable for us using the production config.

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+		],
+	},
+	plugins: [
+		// GUTENBERG_BUNDLE_ANALYZER global variable enables utility that represents bundle content
```
```suggestion
		// BUNDLE_ANALYZER global variable enables utility that represents bundle content
```

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+		process.env.GUTENBERG_BUNDLE_ANALYZER && new BundleAnalyzerPlugin(),
+		// GUTENBERG_LIVE_RELOAD_PORT global variable changes port on which live reload works
+		// when running watch mode.
+		! isProduction && new LiveReloadPlugin( { port: process.env.GUTENBERG_LIVE_RELOAD_PORT || 35729 } ),
```
Same here. Don't see this used anywhere. I'd suggest renaming to:

```suggestion
		! isProduction && new LiveReloadPlugin( { port: process.env.LIVE_RELOAD_PORT || 35729 } ),
```

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+		// GUTENBERG_BUNDLE_ANALYZER global variable enables utility that represents bundle content
+		// as convenient interactive zoomable treemap.
+		process.env.GUTENBERG_BUNDLE_ANALYZER && new BundleAnalyzerPlugin(),
+		// GUTENBERG_LIVE_RELOAD_PORT global variable changes port on which live reload works
```
```suggestion
		// LIVE_RELOAD_PORT global variable changes port on which live reload works
```

nosolosw commented on webpack.config.js:
```diff
@@ -24,166 +23,81 @@ const { dependencies } = require( './package' );
-		],
-	},
-	plugins: [
-		new DefinePlugin( {
```
We seem to have dropped the version flags config in an update/rebase.

gziolo replied:
Yeah, great catch :eagle: :eye:

gziolo replied:
Fixed in https://github.com/WordPress/gutenberg/pull/13814/commits/c1d98342123d0b1430b4c2f4b214bd9cb6df888f.

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+};
+
+if ( ! isProduction ) {
+	// GUTENBERG_DEVTOOL global variable controls how source maps are generated.
```
```suggestion
	// WP_DEVTOOL global variable controls how source maps are generated.
```

nosolosw commented on packages/scripts/config/webpack.config.js:
```diff
@@ -0,0 +1,114 @@
+if ( ! isProduction ) {
+	// GUTENBERG_DEVTOOL global variable controls how source maps are generated.
+	// See: https://webpack.js.org/configuration/devtool/#devtool.
+	config.devtool = process.env.GUTENBERG_DEVTOOL || 'source-map';
```
```suggestion
	config.devtool = process.env.WP_DEVTOOL || 'source-map';
```

oandregal reviewed (commented):
Added a couple of comments I'd like us to address before merging this. We also need to expand the `@wordpress/scripts` README to include the existing defaults (entry, output), how to extend the config, how to use the LiveReload plugin, and the BundleAnalyzer plugin.

Other than those, this is ready and look forward to merge it! Find example of an external project using this at https://github.com/nosolosw/understanding-gutenberg/pull/6).

nosolosw committed:
Replace GUTENBERG with WP in webpack config env variables

Co-Authored-By: gziolo <grzegorz@gziolo.pl>

gziolo committed:
Bring back feature flag to webpack config accidentally removed during merge

gziolo committed:
Add missing dev dependencies for the packages used in webpack config

gziolo committed:
Fix the list of excluded folders for babel-loader

gziolo commented:
> We also need to expand the `@wordpress/scripts` README to include the existing defaults (entry, output

We shouldn't do it in this PR as I treat it as an internal release. We definitely should document it as soon as it is included in `wp-scripts build` and `wp-scripts start`.

> how to extend the config

I would skip it altogether, this is advanced usage which we don't want to promote for the time being. Unless we decide to move this config to its own public package.

> how to use the LiveReload plugin, and the BundleAnalyzer plugin.

Again, I don't think this should be promoted as is. We should keep it internal for some time and figure out whether we can expose it using CLI arguments passed to `wp-scripts`.

nosolosw commented on packages/block-serialization-spec-parser/index.js:
```diff
@@ -1,2 +0,0 @@
-export { parse } from './parser';
```
Do we need the changes in the spec and default parsers? I don't understand why they're necessary. I've undone them, and could `npm install`, `npm run test`, and `npm run dev` without any problem (also tested to load and save an existing post).

gziolo replied:
Yes, I had the same issue as it’s no longer bundled into Gutenberg. There is another parser implementation (more performant one) that is bundled by default. You would have to change it in both PHP and JS files related to parser in wp-blocks. It explodes with umd.

oandregal reviewed (approved):
Let's have this party started :tada:

gziolo commented:
@nosolosw - many thanks for your help making it happen!

gziolo commented:
Kudos to @griffbrad for the initial development of this PR. Great work tackling WordPress externals in particular 🙇

aduth commented on webpack.config.js:
```diff
@@ -24,165 +23,86 @@ const { dependencies } = require( './package' );
-	jquery: 'jQuery',
-	lodash: 'lodash',
-	'lodash-es': 'lodash',
+config.entry = gutenbergPackages.reduce( ( memo, packageName ) => {
```
I think for the purpose it serves, it shouldn't prove to be an issue, but it's generally poor form to be mutating the exported value of a node dependency, as it will impact any other loaded file which imports the same reference.

gziolo replied:
Agreed, it's a non-issue at the moment but it might become one in the future. Fun fact is that at some point of this PR I added an example of how to use such shared config:

https://github.com/WordPress/gutenberg/blob/cc5cb3613607bbe721533f31de7679b28466a3a3/packages/webpack-config/README.md

I used `Object.assign` to extend the default object :) Somehow I didn't follow my own recommendation. I will open a follow-up PR to improve this code to use good practice 👍

gziolo replied:
PR is ready: #14039.
